### PR TITLE
[Weighted Shard Routing] Fail open requests on search shard failures

### PR DESCRIPTION
**Pull request by anshu1106**
### Description
Fail open shard copies request to go to the nodes in weighed away zone. This helps in reducing the number of 5xx responses as well as better shard availability. In other case, where shard search requests return 4xx due to throttling from shard copies in non-weighed away zone, we don't need to fail open. So basically if the request fails due to internal issues in the cluster we fail open for search requests.

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/4735

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by anshu1106** [comment]
> With this change, we will also affect the output of `search_shards` api which will now return all the shards including weight 0 . This might be acceptable behavior, but worth calling out.

Sure. Makes sense.

File: server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java [hunk 1]
```diff
@@ -83,10 +83,18 @@ public class OperationRouting {
         Setting.Property.Dynamic,
         Setting.Property.NodeScope
     );
+
+    public static final Setting<Boolean> WEIGHTED_ROUTING_FAILOPEN_ENABLED = Setting.boolSetting(
+        "cluster.routing.weighted.fail_open",
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java:88]: should we enable it by default as it is already experimental ?

File: server/src/main/java/org/opensearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java [hunk 2]
```diff
@@ -261,24 +262,43 @@ private void onFailure(ShardRouting shardRouting, Exception e) {
...
             if (shardsIt.size() == 0 || shardIndex >= shardsIt.size()) {
                 return null;
             }
             ShardRouting next = shardsIt.get(shardIndex).nextOrNull();
+
+            // This checks if the shard is present in data node with weighted routing weight set to 0,
+            // In such cases we fail open, if shard search request for the shard from other shard copies fail with non
+            // retryable exception.
+            while (next != null && WeightedRoutingHelper.shardInWeighedAwayAZ(next.currentNodeId(), clusterService.state())) {
+                if (WeightedRoutingHelper.isInternalFailure(failure)) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java:275]: I strongly feel we should add allow searches in weighed away nodes, when there are unassigned copies of that shard as well . This would prevent unnecessary 4XX to the user .

File: server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java (hunk 1 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java:88]: We can do it

File: server/src/main/java/org/opensearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java (hunk 2 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java:275]: Yeah makes sense. I can add that check

File: server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    /**
+     * This function checks if the shard copy is present in az with weighted routing weight zero.
+     *
+     * @param nodeId the node with the shard copy
+     * @param clusterState
+     * @return true if the node is in weigh away az ie az with shard routing weight set to zero, else false
+     */
+    public static boolean shardInWeighedAwayAZ(String nodeId, ClusterState clusterState) {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java:49]: Can we create generic methods and not tie to zone/rack

File: server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        DiscoveryNode node = clusterState.nodes().get(nodeId);
+        AtomicBoolean shardInWeighedAwayAZ = new AtomicBoolean(false);
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java:51]: Why AtomicBoolean?

File: server/src/main/java/org/opensearch/action/support/broadcast/TransportBroadcastAction.java [hunk 5]
```diff
@@ -251,6 +252,24 @@ void onOperation(@Nullable ShardRouting shard, final ShardIterator shardIt, int
...
+                        () -> new ParameterizedMessage(
+                            "{}: Fail open executed",
+                            shardToFailOpen.shardId()
+
+                        )
+                    );
+                    break;
+                }
+                nextShard = shardIt.nextOrNull();
+            }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/action/support/broadcast/TransportBroadcastAction.java:272]: Can we create a class called FailOpenRouting and handle the logic there

File: server/src/main/java/org/opensearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java (hunk 2 above)
Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java:275]: Lets avoid redundant handling of logic

File: server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java (hunk 4 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java]: Not needed. Removed

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java [hunk 6]
```diff
@@ -126,18 +131,27 @@ public static ShardRouting findNext(final ShardsIterator shardsIt, Exception fai
...
-    private static boolean canFailOpen(Exception e) {
-        return isInternalFailure(e);
+    private boolean canFailOpen(ShardId shardId) {
+        return isInternalFailure() || hasUnassignedShards(shardId);
+    }
+
+    private boolean hasUnassignedShards(ShardId shardId) {
+        List<ShardRouting> shards = clusterState.routingTable().shardRoutingTable(shardId).shards();
+        for (ShardRouting shardRouting : shards) {
+            if (shardRouting.unassigned()) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:150]: We should add `! shardRouting.active()` here, as it would include unassigned and initializing shards as well. And then rename method as `hasInActiveShardCopies` .

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java [hunk 7]
```diff
@@ -22,29 +22,39 @@
...
-    private static boolean isInternalFailure(Exception e) {
-        return e instanceof NoShardAvailableActionException
-            || e instanceof UnavailableShardsException
-            || e instanceof NodeNotConnectedException
-            || e instanceof NodeDisconnectedException;
+    private boolean isInternalFailure() {
+        return exception instanceof NoShardAvailableActionException
+            || exception instanceof UnavailableShardsException
+            || exception instanceof NodeNotConnectedException
+            || exception instanceof NodeDisconnectedException;
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:57]: `exception` is always going to be `OpenSearchException` and its status will be 5XX for it to be retried. Can we use this instead of hardcoding this exceptions types ?

File: CHANGELOG.md [hunk 8]
```diff
@@ -14,6 +14,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - [Test] Add IAE test for deprecated edgeNGram analyzer name ([#5040](https://github.com/opensearch-project/OpenSearch/pull/5040))
 - Allow mmap to use new JDK-19 preview APIs in Apache Lucene 9.4+ ([#5151](https://github.com/opensearch-project/OpenSearch/pull/5151))
 - Add feature flag for extensions ([#5211](https://github.com/opensearch-project/OpenSearch/pull/5211))
+- Fail open weigh away az in case of failures ([#5072](https://github.com/opensearch-project/OpenSearch/pull/5072))
```
Comment (gbbafna) [inline CHANGELOG.md:17]: nit : `az` not needed.

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+                break;
+            }
+            next = shardsIt.nextOrNull();
+        }
+        return next;
+    }
+
+    private void logFailOpen(ShardId shardID) {
+        logger.info(() -> new ParameterizedMessage("{}: Fail open executed", shardID));
+    }
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:136]: nit : do we need a function here or can inline it directly in `findNext `

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * This class contains logic to find next shard to retry search request in case of failure from other shard copy.
+ * This decides if retryable shard search requests can be tried on shard copies present in data
+ * nodes whose attribute value weight for weighted shard routing is set to zero.
+ */
+
+public class FailOpenRouting {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:35]: Lets write UTs for this class.

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java (hunk 7 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:57]: yeah true, that would make this logic simpler but I am not able to find a way to get status. Do you have any pointer on this?

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java (hunk 9 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:136]: since this is used at multiple place, I think its better to make a function

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java (hunk 7 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:57]: We can use something like below 

```
        assert exception instanceof OpenSearchException;
         return ((OpenSearchException) exception).status().getStatus() / 100 == 5;
```

Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java]: Thanks @gbbafna. Have made the changes to check for 5xx status

**Comment by gbbafna** [comment]
Please fix the failing IT `org.opensearch.search.SearchWeightedRoutingIT.testSearchAggregationWithNetworkDisruption_FailOpenEnabled`

File: server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+     * internal failures
+     *
+     * @param e
+     * @return
+     */
+    public static boolean isInternalFailure(Exception e) {
+        return e instanceof NoShardAvailableActionException
+            || e instanceof UnavailableShardsException
+            || e instanceof NodeNotConnectedException
+            || e instanceof NodeDisconnectedException;
```
[the remark this answers is not in the document]
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingHelper.java]: Update code logic to check for status code

File: server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java [hunk 12]
```diff
@@ -615,7 +615,7 @@ public void testSearchAggregationWithNetworkDisruption_FailOpenEnabled() throws
         for (int i = 0; i < 50; i++) {
             responses[i] = internalCluster().client(nodeMap.get("b").get(0))
                 .prepareSearch("index")
-                .setSize(10)
+                .setSize(20)
```
Comment (gbbafna) [inline server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java:618]: wondering how it fixed the test .

**Review by gbbafna — APPROVED** [review]
LGTM.

File: server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java (hunk 12 above)
Comment (anshu1106) [reply server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java:618]: This was not the only change. Made some changes here as well  https://github.com/opensearch-project/OpenSearch/pull/5072/commits/e76debab289ac087be13f1d25f018a662dc0be1f. 
Basically the search requests were hitting node with network disruption earlier which was causing shard failures

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * *
+     * @return true if exception is due to cluster availability issues
+     */
+    private boolean isInternalFailure() {
+        if (exception instanceof OpenSearchException) {
+            return ((OpenSearchException) exception).status().getStatus() / 100 == 5;
+        }
+        return false;
+    }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:55]: Please lets be more specific here, prefer using something like RestStatus#INTERNAL_SERVER_ERROR

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+    /**
+     * This function checks if the shard is present in data node with weighted routing weight set to 0,
+     * In such cases we fail open, if shard search request for the shard from other shard copies fail with non
+     * retryable exception.
+     *
+     * @param nodeId the node with the shard copy
+     * @return true if the node has attribute value with shard routing weight set to zero, else false
+     */
+    private boolean shardInNodeWithZeroWeightedRouting(String nodeId) {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:65]: nit: Rename to `isWeighedAway`

File: server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java [hunk 15]
```diff
@@ -449,7 +450,8 @@ private void onShardFailure(final int shardIndex, @Nullable SearchShardTarget sh
         // we always add the shard failure for a specific shard instance
         // we do make sure to clean it on a successful response from a shard
         onShardFailure(shardIndex, shard, e);
-        final SearchShardTarget nextShard = shardIt.nextOrNull();
+        SearchShardTarget nextShard = new FailOpenRouting(e, clusterState).findNext(shardIt);
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java:453]: Can we verify that if weights are set only then branch off to `FailureAwareRouting`

File: server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * This class contains logic to find next shard to retry search request in case of failure from other shard copy.
+ * This decides if retryable shard search requests can be tried on shard copies present in data
+ * nodes whose attribute value weight for weighted shard routing is set to zero.
+ */
+
+public class FailOpenRouting {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java:32]: nit: rename to `FailureAwareRouting`

File: server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java (hunk 15 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java:453]: Right now, that is not the case. In all cases, the code flow moves to `FailureAwareRouting`  to fetch next shard copy in case of failures. Should be fine I guess, do you think we should check for weights in action itself.

File: server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java [hunk 17]
```diff
@@ -158,4 +181,617 @@ public void testSearchWithWRRShardRouting() throws IOException {
...
+        }
+
+        NodesStatsResponse nodeStats = client().admin().cluster().prepareNodesStats().execute().actionGet();
+        for (NodeStats stat : nodeStats.getNodes()) {
+            SearchStats.Stats searchStats = stat.getIndices().getSearch().getTotal();
+            if (stat.getNode().isDataNode()) {
+                if (stat.getNode().getId().equals(dataNodeId)) {
+                    Assert.assertTrue(searchStats.getFetchCount() > 0L);
+                    Assert.assertTrue(searchStats.getQueryCount() > 0L);
+                }
```
Comment (Bukhtawar) [inline server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java:551]: Can we assert for the exact count if possible?

File: server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java [hunk 18]
```diff
@@ -317,6 +329,29 @@ public ShardIterator activeInitializingShardsWeightedIt(WeightedRouting weighted
...
+                        ordered.add(iterator.nextOrNull());
+                    }
+                });
+            } catch (IllegalArgumentException e) {
+                // this exception is thrown by {@link onlyNodeSelectorActiveInitializingShardsIt} in case count of shard
+                // copies found is zero
+                logger.debug("no shard copies found for shard id [{}] for node attribute with weight zero", shardId);
+            }
+        }
+
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java:354]: I just realised we aren't maintaining any state across request and hence we might just be implementing an incorrect version of weighted routing.

Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java:354]: State is maintained through across requests using seed https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java#L312

But since the shuffler sits in the IndexShardRoutingTable class, the state would be lost in case of modifications to `IndexShardRoutingTable`

Created a follow up task for this https://github.com/opensearch-project/OpenSearch/issues/5754

File: server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java (hunk 17 above)
Comment (anshu1106) [reply server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java:551]: Its difficult to get same distribution of shards on every test run, hence we can't assert on actual number. But we have a [task](https://github.com/opensearch-project/OpenSearch/issues/5037) opened to get fail open stats which can help us track this closer.

File: CHANGELOG.md [hunk 19]
```diff
@@ -15,6 +15,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - [Test] Add IAE test for deprecated edgeNGram analyzer name ([#5040](https://github.com/opensearch-project/OpenSearch/pull/5040))
 - Allow mmap to use new JDK-19 preview APIs in Apache Lucene 9.4+ ([#5151](https://github.com/opensearch-project/OpenSearch/pull/5151))
 - Add feature flag for extensions ([#5211](https://github.com/opensearch-project/OpenSearch/pull/5211))
+- Fail open weigh away in case of failures ([#5072](https://github.com/opensearch-project/OpenSearch/pull/5072))
```
Comment (Bukhtawar) [inline CHANGELOG.md:18]: Please reword this as 
```
Support to fail open requests on search shard failures 
```

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    /**
+     * *
+     * @return true if exception is due to cluster availability issues
+     */
+    private boolean isInternalFailure() {
+        if (exception instanceof OpenSearchException) {
+            // checking for 5xx failures
+            return (((OpenSearchException) exception).status().getStatus() / 100) * 100 == RestStatus.INTERNAL_SERVER_ERROR.getStatus();
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:54]: 1. Create a Immutable list of exhaustive 5xx codes 
2. Compare the status against items on that list
3. Remove division/mul by 100

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+     * This function returns next shard copy to retry search request in case of failure from previous copy returned
+     * by the iterator. It has the logic to fail open ie request shard copies present in nodes with weighted shard
+     * routing weight set to zero
+     *
+     * @param shardIt Shard Iterator containing order in which shard copies for a shard need to be requested
+     * @return the next shard copy
+     */
+    public SearchShardTarget findNext(final SearchShardIterator shardIt) {
+        SearchShardTarget next = shardIt.nextOrNull();
+        while (next != null && isWeighedAway(next.getNodeId())) {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:101]: Can we simply check if there are shard routing that are not a part of the original iterator(without appended shards with weights zero) which can be used to retry/execute requests. The reason is there could be a race between when the request was first executed and when the request got retried(typically during concurrent failover requests)

T0 -> Weights(2:1:0)
T1 -> Requests failed on zone1 and then zone2
T2 -> Weights(0:1:2)
T3 -> Retry on third zone will be computed but this check(`isWeighedAway`) will fail this and prevent this request from executing successfully

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:101]: I realise that this check is put in place to avoid the shuffler from picking up zones that are weighed away, but can we do something to avoid independent check between when we add it to the iterator to when we actually consume it which is here

Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:101]: yeah, this check is for weigh away zones, to avoid iterator to pick shards from weigh away az nodes in cases we don't want to fail open.




> but can we do something to avoid independent check between when we add it to the iterator to when we actually consume it which is here

I think most of the cases will have only one check but in worst case scenario where there is a need to fail open like cases we may have to iterate through all the shards and do `isWeighedAway` check. This won't be a very costly operation since size of the list is fixed.
Even if we precompute, weights can change in between which will require one more check.

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java (hunk 20 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:54]: We can create a list of 5xx codes but that will require regular update. Did this since we are interested in all OpenSearchException with 5xx status code. 

Any specific reason I have missed for which you are suggesting this?

File: server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java [hunk 22]
```diff
@@ -317,6 +330,29 @@ public ShardIterator activeInitializingShardsWeightedIt(WeightedRouting weighted
...
```
[the remark this answers is not in the document]
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java]: Updated code to return distinct shards

File: server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java [hunk 23]
```diff
@@ -352,8 +354,8 @@ public ShardIterator activeInitializingShardsWeightedIt(
                 logger.debug("no shard copies found for shard id [{}] for node attribute with weight zero", shardId);
             }
         }
-
-        return new PlainShardIterator(shardId, ordered);
+        orderedListWithDistinctShards = new ArrayList<>(new LinkedHashSet<>(ordered));
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java:357]: Suggest
```
ordered.stream().distinct().collect(Collectors.toList())
```

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+        RestStatus.SERVICE_UNAVAILABLE,
+        RestStatus.GATEWAY_TIMEOUT,
+        RestStatus.HTTP_VERSION_NOT_SUPPORTED,
+        RestStatus.INSUFFICIENT_STORAGE
+    );
+
+    public FailAwareWeightedRouting(Exception e, ClusterState clusterState) {
+        this.exception = e;
+        this.clusterState = clusterState;
+    }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:55]: Please use ClusterService in the constructor so that it could be used as a singleton instance and move exception to the method `findNext` and log the same exception

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+                logFailOpen(next.shardId());
+                break;
+            }
+            next = shardsIt.nextOrNull();
+        }
+        return next;
+    }
+
+    private void logFailOpen(ShardId shardID) {
+        logger.info(() -> new ParameterizedMessage("{}: Fail open executed", shardID));
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:143]: nit: will prefer this inline

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java (hunk 24 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:55]: I think we can't make `FailAwareWeightedRouting` singleton since even cluster state can change between requests

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:52]: Yeah bad edit, I initially meant ClusterService for singleton but forgot to remove.

**Comment by Bukhtawar** [comment]
Test failures : https://github.com/opensearch-project/OpenSearch/issues/5766

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java (hunk 24 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java]: Got it. I have made `FailAwareWeightedRouting` singleton but we can't pass `ClusterService` in the constructor since `ClusterService` instance is not available in `AbstractSearchAsyncAction`.

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+        RestStatus.INTERNAL_SERVER_ERROR,
+        RestStatus.BAD_GATEWAY,
+        RestStatus.SERVICE_UNAVAILABLE,
+        RestStatus.GATEWAY_TIMEOUT
+    );
+
+    public static FailAwareWeightedRouting getInstance() {
+        return INSTANCE;
+
+    }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:48]: I don't think this is the most apt way to create singletons, you could achieve it by a static property
```
    public static final FailAwareWeightedRouting INSTANCE = new FailAwareWeightedRouting();
```

File: server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+     * @param shardsIt Shard Iterator containing order in which shard copies for a shard need to be requested
+     * @return the next shard copy
+     */
+    public ShardRouting findNext(final ShardsIterator shardsIt, ClusterState clusterState, Exception exception) {
+        ShardRouting next = shardsIt.nextOrNull();
+
+        while (next != null && isWeighedAway(next.currentNodeId(), clusterState)) {
+            ShardRouting nextShard = next;
+            if (canFailOpen(nextShard.shardId(), exception, clusterState)) {
+                logger.info(() -> new ParameterizedMessage("{}: Fail open executed due to exception {}", nextShard.shardId(), exception));
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java:130]: nitL incorrect usage of logger with exception

**Review by Bukhtawar — APPROVED** [review]
